rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,578 @@
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(dst),
+                err => Err(SymCryptError::from(err)),
+            }
+        }
+    }
+}
+
+impl RsaKey for RsaKeyPair {
+    fn inner(&self) -> symcrypt_sys::PSYMCRYPT_RSAKEY {
+        self.inner
+    }
+
+    fn key_usage(&self) -> RsaKeyUsage {
+        self.key_usage
+    }
+}
+
+impl RsaKey for RsaPublicKey {
+    fn inner(&self) -> symcrypt_sys::PSYMCRYPT_RSAKEY {
+        self.inner
+    }
+
+    fn key_usage(&self) -> RsaKeyUsage {
+        self.key_usage
+    }
+}
+
+impl RsaKeyPair {
+    /// `generate_new()` generates a random Rsa key based on the provided parameters.
+    ///
+    /// `n_bits_mod` represents the desired bit length of the Rsa Key.
+    ///
+    /// `pub_exp` takes in an `Option<u64>` that is the public exponent. If `None` is provided, the default `2^16 +1` will be used.
+    ///  
+    /// `key_usage` will indicate if this key will be used for [`RsaKeyUsage::Sign`] or [`RsaKeyUsage::Encrypt`]
+    pub fn generate_new(
+        n_bits_mod: u32,
+        pub_exp: Option<&[u8]>,
+        key_usage: RsaKeyUsage,
+        num_format: NumberFormat,
+    ) -> Result<Self, SymCryptError> {
+        let rsa_key = allocate_rsa(2, n_bits_mod); // !Review: @Phil has mentioned that there is no scenario that we'd want to ONLY generate a public key, @Crypto folks is this the case?
+        let u64_pub_exp = match pub_exp {
+            Some(exp) => RsaKeyPair::load_msb_first_u64(exp)?,
+            None => DEFAULT_PUBLIC_EXPONENT, // if no public exponent is provided, use the default ( 2^16 + 1 )
+        };
+        unsafe {
+            // SAFETY: FFI calls
+            // No flags specified for SymCryptRsakeyAllocate
+            match symcrypt_sys::SymCryptRsakeyGenerate(
+                rsa_key,
+                [u64_pub_exp].as_ptr(), 
+                1,
+                key_usage.to_flag(),
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(RsaKeyPair {
+                    inner: rsa_key,
+                    key_usage: key_usage,
+                    number_format: num_format,
+                }),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// `set_key_pair()` sets both the public and private key information onto [`RsaKeyPair`].
+    ///
+    /// `modulus_buffer` takes in a reference to a byte array that contains the modulus of the Rsa key.
+    ///
+    /// `pub_exp` takes in a `&[u8]` that is the public exponent represented by an array of bytes.
+    ///
+    /// `p` takes in reference to a byte array that contains the first prime.
+    ///
+    /// `q` takes in reference to a byte array that contains the second prime.
+    ///
+    /// `num_format` takes in a [`NumberFormat`] that specifies either [`NumberFormat::LSB`] or [`NumberFormat::MSB`] which must match ALL inputs.
+    pub fn set_key_pair(
+        n_bits_mod: u32,
```

Comment:
>  n_bits_mod: u32, [](http://example.com/codeflow?start=6&length=18)

Seems like this should be derived from length of modulus buffer?

Reply:
@mlindgren  @samuel-lee-msft do we know who derives the bit size? what if the key that we're importing has `len()` of 2047